test/dynamo/test_backward_higher_order_ops.py

```diff
@@ -0,0 +1,313 @@
+
+        copy_ = torch.ops.aten.copy_.default(new_empty_strided, getitem);  new_empty_strided = None
+
+        mul = getitem * getitem;  getitem = None
+
+        call_hook = torch__dynamo__trace_wrapped_higher_order_op__assert_meta(mul, (2,), (1,), torch.float32);  mul = None
+
+        new_empty_strided_1 = torch.ops.aten.new_empty_strided.default(call_hook, [2], [1], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
+
+        copy__1 = torch.ops.aten.copy_.default(new_empty_strided_1, call_hook);  new_empty_strided_1 = call_hook = None
+        return (copy_, copy__1)
+"""
+            self.assertExpectedInline(actual, expected)
+
+            out = fn(x, y)
+            out.backward(grad_out)
+            self.assertEqual(obj.counter, 2)
+
+            out = fn(x, y)
+            out.backward(grad_out)
+            self.assertEqual(obj.counter, 3)
+            graph = None
+
+    def test_invoke_in_pt2_compiled_autograd_graph_breaks(self):
+        def _graph_breaking_fn(x):
+            print("Boo!")
+            return _multiply(x)
+
+        def _graph_break_invoke(grad):
+            return trace_wrapped(grad, fn=_graph_breaking_fn)
+
+        def compiler_fn(gm):
+            return torch.compile(gm, backend="inductor", fullgraph=True, dynamic=True)
+
+        for backend in ["eager", "aot_eager", "inductor"]:
+            torch._dynamo.reset()
+            x = torch.tensor([0.5, 0.5], requires_grad=True)
+            y = torch.tensor([0.5, 0.5], requires_grad=True)
+
+            def fn(x, y):
+                x.register_hook(_graph_break_invoke)
+                return x + y
+
+            fn = torch._dynamo.optimize(backend, nopython=True)(fn)
+            out = fn(x, y)
+            grad_out = torch.tensor([2.0, 2.0])
+            with self.assertRaisesRegex(
+                torch._dynamo.exc.Unsupported,
+                "print",
+            ):
+                with compiled_autograd.enable(compiler_fn):
+                    out.backward(grad_out)
+
+            graph = None
+
+    def test_invoke_in_pt2_compiled_autograd_mutates_tensors(self):
+        def _tensor_mutating_stride(x):
+            return torch.as_strided(x, size=x.shape, stride=(1, 3))
+
+        def _tensor_mutating_size(x):
+            return torch.randn([9, 1])
+
+        def _tensor_mutating_dtype(x):
+            return x.to(dtype=torch.float64)
+
+        def _tensor_mutating_dtype(x):
+            return x.to(dtype=torch.float64)
+
+        def _returning_not_tensor(x):
+            return [x, x]
+
+        for problem, mut_fn in [
+            ("non-tensor", _returning_not_tensor),
+            ("stride", _tensor_mutating_stride),
+            ("size", _tensor_mutating_size),
+            ("dtype", _tensor_mutating_dtype),
+        ]:
+
+            def _graph_break_invoke(grad):
```

Comment:
This technically won't cause a graph break (as its name implies) right - it'll cause a hard error in the backward?

Reply:
Also, it would be nice to have another test for aliasing, where a hook returns a view of the input. (which is "wrong" because our fake tensor rule for the wrap higher order op assumes that the output of the hook never aliases the input).

Reply:
It's a graph break, which turns into a hard error.